test: add windows UT support and multiple tests and VS projects

peluse opened the pull request:
Also misc modifications to support windows compilation. Not all
tests are included in this patch, just the required framework,
two sets of tests, blk_nblock and blk_non_zero, and several but
not all VS project files for the tests

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/nvml/819)
<!-- Reviewable:end -->

marcinslusarz commented:
This approach is unmaintainable - we extend the framework very frequently, so the requirement to update it twice for different languages/platforms is infeasible. Instead of adding 2nd unportable testing framework we need to replace the existing framework by a portable one, e.g. written in Python.
Even now your framework code is out-of-sync with bash implementation (e.g. for several months there is no LOCAL_FS_DIR variable).

As for the C files, you need to split them into separate commits and describe why they are needed. I think most of them are not needed and are the result of basing this work on top of previous version of the Windows patch. (setvbuf, UT_ASSERTeq_rt; open change is also fishy, but I don't know why you did it; struct stat change should be abstracted).

Why do you need separate match files for Windows? Current files can be easily adapted to support both systems.

peluse commented:
- The approach was discussed at length with Krzy and Andy and your concerns were part of those discussions and they are valid. However, the consensus was that it is maintainable and that the effort needed to rework everything into a single language was not worth it,  Andy is on site this week, please get with him to discuss this point.
- Yes, its out of date now because this is the initial set of powershell scripts. Once this large patch is merged it will not be that much work to keep them in sync, this has been in the making for weeks and will just continue to grow if I attempt to keep it in sync. What is on master now (including this patch) does not represent a fully workable windows solution obviously.  
- Why do the C file changes needed to be in a separate commit? The changes there are needed to support execution of the powershell scripts, all of these changes are tied together
- Different match files are needed because not all of the tests have the same output. For example, only two test dirs have been ported so far and at least 2 of those individual tests have different output than the linux versions because of the inability to test very large files under windows. Most of them should be the same though.
- setvbuf change-  do you have a specific question? The comment states it clearly I think, the windows clib doesn't support that call so we use a different one for windows.

<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> This approach is unmaintainable - we extend the framework very frequently, so the requirement to update it twice for different languages/platforms is infeasible. Instead of adding 2nd unportable testing framework we need to replace the existing framework by a portable one, e.g. written in Python.
> 
> Even now your framework code is out-of-sync with bash implementation (e.g. for several months there is no LOCAL_FS_DIR variable).
> 
> 
> 
> As for the C files, you need to split them into separate commits and describe why they are needed. I think most of them are not needed and are the result of basing this work on top of previous version of the Windows patch. (setvbuf, UT_ASSERTeq_rt; open change is also fishy, but I don't know why you did it; struct stat change should be abstracted).
> 
> 
> 
> Why do you need separate match files for Windows? Current files can be easily adapted to support both systems.
> 
> </details>


---

Review status: 0 of 111 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819#-:-KHtSzIFCbvRJgt0oM9q:-60943083)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Reviewed 111 of 111 files at r1.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/pmem/nvml/819#-KHusBz2UVmZH06UkBfu:-KHusBz3vLzEsttMDIwc:13982444) (no related file):*
Some general comments:

1) Please, don't check for `WIN32`, but `_WIN32`.

2) I don't like the idea of having separate `*.match` files for Linux and Windows. 
2.1) Handling different directory separators could be somehow addressed in `match` script, I believe.
Optionally, we can do some pre-processing of the output files and replace backslashes with slashes. Or, we can modify out.c/ut.c to make sure that all the file paths are always printed with "/" as dir separator.

2.2) If some scenarios cannot be run on Windows (i.e. those libpmemblk tests that create huge sparse files), we should rather split the existing `TEST*` files to make sure that all the scenarios from given script can be executed on both platforms. Then, the match file could be shared by both OS-es.
Scenarios that cannot work on Windows would be in a separate TESTx script that will have no Windows TESTx.ps1  counterpart.

---

*[src/NVML.sln, line 121 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv6VsupZUrVEsgOCsP:-KHv6VsupZUrVEsgOCsQ:2045241019) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/NVML.sln#L121)):*
> ```
> Global
> 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
> 		Debug|ARM = Debug|ARM
> ```

Please, remove ARM/x86 configurations.

---

*[src/NVML.sln, line 147 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv6IvsbLqwIl93CLJ1:-KHv6IvsbLqwIl93CLJ2:-914265319) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/NVML.sln#L147)):*
> ```
> 		{9E9E3D25-2139-4A5D-9200-18148DDEAD45}.Static-Debug|x64.Build.0 = Static-Debug|x64
> 		{9E9E3D25-2139-4A5D-9200-18148DDEAD45}.Static-Debug|x86.ActiveCfg = Static-Debug|x64
> 		{9E9E3D25-2139-4A5D-9200-18148DDEAD45}.Static-Release|ARM.ActiveCfg = Static-Debug|x64
> ```

ARM and x86 configurations are not needed.

---

*[src/common/liblinux.vcxproj.filters, line 18 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv89obdLlgmSfAJBEg:-KHv89oc99Zh57VDNoWu:705013469) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/common/liblinux.vcxproj.filters#L18)):*
> ```XML
>   </ItemGroup>
>   <ItemGroup>
>     <ClInclude Include="..\include\sys\file.h">
> ```

Fix the path to this file. Otherwise, VS cannot find it and rebuilds liblinux each time.

---

*[src/common/out.c, line 243 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuMgyCHh-4VjEkwalw:-KHuMgyCHh-4VjEkwalx:440875221) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/common/out.c#L243)):*
> ```C
> #else
> 		/* setlinebuf() is not available on Windows, use no buffering */
> 		setvbuf(Out_fp, NULL, _IONBF, 0);
> ```

Looks like on Windows, it is not allowed to pass 0 as the buffer size, and this is why the previous version crashes (sorry for not validationg this).  So, instead of disabling buffering, perhaps we should use *BUFSIZ* as the buffer size?   On Linux, it seems to be ignored anyway (if *buf* is NULL), but on Windows, the buffer of given size is implicitly allocated (which is what we want).
IOW, either we should  simply replace all the calls to `setlinebuf(fp)` with `setvbuf(fp, NULL, _IOLBF, BUFSIZ)`, or define `setlinebuf()` macro (platform.h)  this way.  IMHO, there is no need to disable buffering completely.

---

*[src/common/util_windows.c, line 54 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHurscmzoexa8ETMP56:-KHurscmzoexa8ETMP57:-792471532) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/common/util_windows.c#L54)):*
> ```C
> {
> 	LOG(4, "hint not supported on windows");
> 	return NULL;
> ```

What's the reason of this change?

---

*[src/test/RUNTESTS, line 107 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv8nUEkl5zgABaVrmo:-KHv8nUEkl5zgABaVrmp:1815473528) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/RUNTESTS#L107)):*
> ```
> 	[ "$builds" = all ] && builds="debug nondebug static-debug static-nondebug"
> 	runscripts=$testfile
> 	[ "$runscripts" = all ] && runscripts=`ls -1 TEST* | sort -V`
> ```

I'd rather filter out *.ps1 files here, i.e.:
`... | grep -v "*.ps1" | sort -V`

---

*[src/test/testconfig.ps1, line 1 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv66ICiA-lOkJkJhPj:-KHv66IDfpeB_zdmBMcB:56065803) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/testconfig.ps1#L1)):*
> ```PowerShell
> #
> ```

This file should not be in repo.

---

*[src/test/testconfig.ps1.example, line 28 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv61C3Ck2iZMLhQqh1:-KHv61C3Ck2iZMLhQqh2:1317411242) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/testconfig.ps1.example#L28)):*
> ```
> # files) set the path to a directory on a normal file system here.
> #
> sv -name NON_PMEM_FS_DIR '\tmp'
> ```

\temp ?

---

*[src/test/blk_nblock/blk_nblock.c, line 65 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHux6C-jMc4eHq5ZKeZ:-KHux6C-jMc4eHq5ZKe_:-576265481) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_nblock/blk_nblock.c#L65)):*
> ```C
> 			UT_OUT("%s: block size %zu usable blocks: %zu",
> 					fname, bsize, pmemblk_nblock(handle));
> 			UT_ASSERTeq_rt(pmemblk_bsize(handle), bsize);
> ```

IMHO, this change is not necessary. The problem with UT_ASSERT/T_COMPILE_ERROR_ON macros should be solved in a similar way as in util.h.
Same applies to other tests.

---

*[src/test/blk_nblock/TEST0.ps1, line 76 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv0I3-4nhWslFktkxP:-KHv0I3-4nhWslFktkxQ:837283617) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_nblock/TEST0.ps1#L76)):*
> ```PowerShell
> # provided on the linux side
> #
> #create_holey_file 536576 $DIR\testfile3.512
> ```

What about splitting this filre and moving all the commented out scenarios into a new TEST script?

---

*[src/test/blk_non_zero/blk_non_zero.c, line 97 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuxTUXspueM2rvj3Iw:-KHuxTUXspueM2rvj3Ix:974537893) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_non_zero/blk_non_zero.c#L97)):*
> ```C
> 
> #ifndef WIN32
> 	struct stat stbuf;
> ```

See util.h - I think we need to do the same here.

---

*[src/test/unittest/unittest.ps1, line 40 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv5OXRaUr37pPHxvbY:-KHv5OXSn0ZYc-260IWh:-1704648906) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/unittest/unittest.ps1#L40)):*
> ```PowerShell
> function isDir {
>     if ((Get-Item $args[0] -ErrorAction SilentlyContinue) -is [System.IO.DirectoryInfo]) {
>         return $true			
> ```

Would be nice to unify  indentation - either tabs or spaces, and fix trailing white space.

---

*[src/windows/include/platform.h, line 71 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuXD3fW07fEExdzUAw:-KHuXD3fW07fEExdzUAx:-122736716) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L71)):*
> ```objc
> #define __func__ __FUNCTION__
> #define __typeof__ decltype
> #define restrict __restrict
> ```

Not required anymore.  If *restrict* is still used anywhere, it should be replaced with *__restrict*.

---

*[src/windows/include/platform.h, line 78 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHueUBzF_iquaI-ymAb:-KHueUC-JMKzWg7IFwtP:-1370110660) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L78)):*
> ```objc
> 
> /* XXX - dummy */
> typedef int sigset_t;
> ```

Not needed. At least, not yet...

---

*[src/windows/include/platform.h, line 79 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuXUn8yOOHS3FzW45D:-KHuXUn9KZMqWuU0eJX4:-287849673) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L79)):*
> ```objc
> /* XXX - dummy */
> typedef int sigset_t;
> typedef int DIR;
> ```

All the functions/macros depending on *DIR* are commented out for Windows, so this is not required.
Moreover, those functions/macros are not used in any test, even on Linux.

---

*[src/windows/include/platform.h, line 80 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuUC9Ae0WdodgvGufD:-KHuUC9B9m-J07jJ_8he:1724326011) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L80)):*
> ```objc
> typedef int sigset_t;
> typedef int DIR;
> #define setlinebuf(o)
> ```

See comment in out.c

---

*[src/windows/include/platform.h, line 81 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuplINV4slhbE_QNwM:-KHuplINV4slhbE_QNwN:-1442546860) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L81)):*
> ```objc
> typedef int DIR;
> #define setlinebuf(o)
> #define ut_register_sighandlers(o)
> ```

I think it should rather be inline function (to get rid of some compiler warnings).
Also, as the next step, we need to add `ut_backtrace.c` to `libut` with proper implementation of `ut_dump_backtrace().`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/pmem/nvml/819#-KHusBz2UVmZH06UkBfu:-KHvBpNfrwYQg35sU3aK:-31156084) (no related file):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Some general comments:
> 
> 1) Please, don't check for `WIN32`, but `_WIN32`.
> 
> 2) I don't like the idea of having separate `*.match` files for Linux and Windows. 
> 2.1) Handling different directory separators could be somehow addressed in `match` script, I believe.
> Optionally, we can do some pre-processing of the output files and replace backslashes with slashes. Or, we can modify out.c/ut.c to make sure that all the file paths are always printed with "/" as dir separator.
> 
> 2.2) If some scenarios cannot be run on Windows (i.e. those libpmemblk tests that create huge sparse files), we should rather split the existing `TEST*` files to make sure that all the scenarios from given script can be executed on both platforms. Then, the match file could be shared by both OS-es.
> Scenarios that cannot work on Windows would be in a separate TESTx script that will have no Windows TESTx.ps1  counterpart.
> 
> </details>
1) Yeah, yeah my eyesight must be going :) I'll change those.
2.1) Yes, we can do that but I didn't because of 2.2
2.2) Hmm, didn't think about that. I will experiment with one case and see if any other rocks are uncovered when trying to do it that way... 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[src/common/out.c, line 243 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuMgyCHh-4VjEkwalw:-KHwK9DFRI-bQLdWR7gv:-420195379) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/common/out.c#L243)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Looks like on Windows, it is not allowed to pass 0 as the buffer size, and this is why the previous version crashes (sorry for not validationg this).  So, instead of disabling buffering, perhaps we should use *BUFSIZ* as the buffer size?   On Linux, it seems to be ignored anyway (if *buf* is NULL), but on Windows, the buffer of given size is implicitly allocated (which is what we want).
> IOW, either we should  simply replace all the calls to `setlinebuf(fp)` with `setvbuf(fp, NULL, _IOLBF, BUFSIZ)`, or define `setlinebuf()` macro (platform.h)  this way.  IMHO, there is no need to disable buffering completely.
> </details>
Sure, I will try that thanks!

---

*[src/common/util_windows.c, line 54 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHurscmzoexa8ETMP56:-KHwKMLcNvEhZOAzT0BV:751104706) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/common/util_windows.c#L54)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> What's the reason of this change?</details>
The base code was returning FAILED if hint isn't supported (instead of NULL)

---

*[src/test/RUNTESTS, line 107 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv8nUEkl5zgABaVrmo:-KHwKYazFNvxwVQxVxXt:592195330) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/RUNTESTS#L107)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I'd rather filter out *.ps1 files here, i.e.:
> `... | grep -v "*.ps1" | sort -V`</details>
OK, didn't know how to do that :)

---

*[src/test/testconfig.ps1.example, line 28 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv61C3Ck2iZMLhQqh1:-KHwL1UAZmJvTBWOVklG:295479925) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/testconfig.ps1.example#L28)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> \temp ?</details>
sure :)

---

*[src/test/blk_nblock/blk_nblock.c, line 65 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHux6C-jMc4eHq5ZKeZ:-KHwLAyFX-1S-jTyokgI:-1860596080) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_nblock/blk_nblock.c#L65)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> IMHO, this change is not necessary. The problem with UT_ASSERT/T_COMPILE_ERROR_ON macros should be solved in a similar way as in util.h.
> Same applies to other tests.</details>
OK, I'll go look and see what this is but if you add some more detail here that's be great.  These new macros were introduced after I started and I still had more work to do so switched to the ones that we were using that don't casus e VS to throw up

---

*[src/test/blk_nblock/TEST0.ps1, line 76 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv0I3-4nhWslFktkxP:-KHwLSA0nX63k6y1unJc:650132796) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_nblock/TEST0.ps1#L76)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> What about splitting this filre and moving all the commented out scenarios into a new TEST script?</details>
I think it'd be better just to delete the comments, a new TEST script wouldn't do anything if I understand what you are suggesting

---

*[src/test/blk_non_zero/blk_non_zero.c, line 97 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuxTUXspueM2rvj3Iw:-KHwLdtPTJHvnnlSY5sN:-2103092626) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_non_zero/blk_non_zero.c#L97)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> See util.h - I think we need to do the same here.</details>
OK will look

---

*[src/test/unittest/unittest.ps1, line 40 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv5OXRaUr37pPHxvbY:-KHwLlMNxi-sXRoJN5X1:-290891961) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/unittest/unittest.ps1#L40)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Would be nice to unify  indentation - either tabs or spaces, and fix trailing white space.</details>
yeah, my editor was setup for spaces instead of tabs at first so everything I touched got converted. Then a bunch of tests failed so I changed it to use tabs instead of spaces and remove trailing white space but I need to touch any files that are affected.  What is our tab count?

---

*[src/windows/include/platform.h, line 71 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuXD3fW07fEExdzUAw:-KHwMCGydzXUvdgrSW2g:-1648492793) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L71)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Not required anymore.  If *restrict* is still used anywhere, it should be replaced with *__restrict*.</details>
it is used somewhere, I will remove and find out then replace (I added this because of failure to compile)

---

*[src/windows/include/platform.h, line 78 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHueUBzF_iquaI-ymAb:-KHwMLi3vaWWxf0PCIzj:907650044) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L78)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Not needed. At least, not yet...</details>
which one?  If I added it, it was required to compile (unless copy n paste error or something).  If not then it was already there

---

*[src/windows/include/platform.h, line 79 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuXUn8yOOHS3FzW45D:-KHwMTuT9vhvFNrZ1l8b:-291444863) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L79)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> All the functions/macros depending on *DIR* are commented out for Windows, so this is not required.
> Moreover, those functions/macros are not used in any test, even on Linux.</details>
Again, I added it because of a failure to compile so something got missed somewhere.  I'll take it out and see who complains

---

*[src/windows/include/platform.h, line 81 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuplINV4slhbE_QNwM:-KHwMiU9oP6Oddtf84FL:1468646510) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L81)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> I think it should rather be inline function (to get rid of some compiler warnings).
> Also, as the next step, we need to add `ut_backtrace.c` to `libut` with proper implementation of `ut_dump_backtrace().`</details>
will take a look

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[src/NVML.sln, line 121 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv6VsupZUrVEsgOCsP:-KHwR-nrpjOfchLbIBW-:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/NVML.sln#L121)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Please, remove ARM/x86 configurations.</details>
Done.

---

*[src/NVML.sln, line 147 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv6IvsbLqwIl93CLJ1:-KHwR19-Ae7B9yDGhl-_:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/NVML.sln#L147)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> ARM and x86 configurations are not needed.</details>
Done.

---

*[src/common/liblinux.vcxproj.filters, line 18 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv89obdLlgmSfAJBEg:-KHwR2wpUli3Fc_DoJ9y:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/common/liblinux.vcxproj.filters#L18)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> Fix the path to this file. Otherwise, VS cannot find it and rebuilds liblinux each time.</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---

*[src/NVML.sln, line 147 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv6IvsbLqwIl93CLJ1:-KHwRH3ex-ldUt1fUiEY:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/NVML.sln#L147)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> Done.</details>
Done.

---

*[src/common/util_windows.c, line 54 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHurscmzoexa8ETMP56:-KHwRW4_NgFLjK3XLiTi:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/common/util_windows.c#L54)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> The base code was returning FAILED if hint isn't supported (instead of NULL)</details>
Done.

---

*[src/test/testconfig.ps1, line 1 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv66ICiA-lOkJkJhPj:-KHwRxeYSP7wxs7rtQit:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/testconfig.ps1#L1)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> This file should not be in repo.</details>
Done.

---

*[src/test/testconfig.ps1.example, line 28 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv61C3Ck2iZMLhQqh1:-KHwS6dTF1DRAZZZtLWa:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/testconfig.ps1.example#L28)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> sure :)</details>
Done.

---

*[src/test/blk_nblock/TEST0.ps1, line 76 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv0I3-4nhWslFktkxP:-KHwSfJigq9NXBWUXx8o:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_nblock/TEST0.ps1#L76)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> I think it'd be better just to delete the comments, a new TEST script wouldn't do anything if I understand what you are suggesting</details>
Done.

---

*[src/test/blk_non_zero/blk_non_zero.c, line 97 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuxTUXspueM2rvj3Iw:-KHwShfw79WTVmKRlTmM:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_non_zero/blk_non_zero.c#L97)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> OK will look</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

ldorau commented:
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[src/test/blk_non_zero/TEST4.ps1, line 37 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHxc31O3vkf3Rs0VB9x:-KHxc31PFb0AFzS8_7wO:1659063552) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_non_zero/TEST4.ps1#L37)):*
> ```PowerShell
> $Env:UNITTEST_NAME = "blk_non_zero\TEST4"
> $Env:UNITTEST_NUM = "4"
> # Pl TODO:  bash has a few calls to tools that we don't have on
> ```

We use 'XXX' instead of 'TODO' or 'TBD' in NVML - please replace them:

src/test/blk_non_zero/TEST4.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
src/test/blk_non_zero/TEST5.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
src/test/blk_non_zero/TEST6.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
src/test/blk_non_zero/TEST7.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
src/test/blk_non_zero/TEST8.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
src/test/blk_non_zero/TEST9.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on

---

*[src/test/blk_non_zero/TEST4.ps1, line 39 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHxctF2DHfphmBV44Ic:-KHxctF2DHfphmBV44Id:-124101310) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_non_zero/TEST4.ps1#L39)):*
> ```PowerShell
> # Pl TODO:  bash has a few calls to tools that we don't have on
> # windows (yet) that set PMEM_IS_PMEM and NON_PMEM_IS_PMEM based
> # on their outpute
> ```

The same typo ('outpute') in the following places:

src/test/blk_nblock/TEST0.ps1:38: on their outpute
src/test/blk_nblock/TEST1.ps1:38: on their outpute
src/test/blk_nblock/TEST2.ps1:38: on their outpute
src/test/blk_non_zero/TEST0.ps1:39: on their outpute
src/test/blk_non_zero/TEST1.ps1:39: on their outpute
src/test/blk_non_zero/TEST2.ps1:39: on their outpute
src/test/blk_non_zero/TEST3.ps1:39: on their outpute
src/test/blk_non_zero/TEST4.ps1:39: on their outpute
src/test/blk_non_zero/TEST5.ps1:39: on their outpute
src/test/blk_non_zero/TEST6.ps1:39: on their outpute
src/test/blk_non_zero/TEST7.ps1:39: on their outpute
src/test/blk_non_zero/TEST8.ps1:39: on their outpute
src/test/blk_non_zero/TEST9.ps1:39: on their outpute

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: all files reviewed at latest revision, 20 unresolved discussions.

---

*[src/test/blk_non_zero/TEST4.ps1, line 37 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHxc31O3vkf3Rs0VB9x:-KHycmnv564w-MTZTdO6:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_non_zero/TEST4.ps1#L37)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary>
> We use 'XXX' instead of 'TODO' or 'TBD' in NVML - please replace them:
> 
> src/test/blk_non_zero/TEST4.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
> src/test/blk_non_zero/TEST5.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
> src/test/blk_non_zero/TEST6.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
> src/test/blk_non_zero/TEST7.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
> src/test/blk_non_zero/TEST8.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on
> src/test/blk_non_zero/TEST9.ps1:37:# Pl TODO:  bash has a few calls to tools that we don't have on</details>
Done.

---

*[src/test/blk_non_zero/TEST4.ps1, line 39 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHxctF2DHfphmBV44Ic:-KHyd8ToF2cj8BFuABs4:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_non_zero/TEST4.ps1#L39)):*
<details><summary><i>Previously, ldorau (Lukasz Dorau) wrote…</i></summary>
> The same typo ('outpute') in the following places:
> 
> src/test/blk_nblock/TEST0.ps1:38: on their outpute
> src/test/blk_nblock/TEST1.ps1:38: on their outpute
> src/test/blk_nblock/TEST2.ps1:38: on their outpute
> src/test/blk_non_zero/TEST0.ps1:39: on their outpute
> src/test/blk_non_zero/TEST1.ps1:39: on their outpute
> src/test/blk_non_zero/TEST2.ps1:39: on their outpute
> src/test/blk_non_zero/TEST3.ps1:39: on their outpute
> src/test/blk_non_zero/TEST4.ps1:39: on their outpute
> src/test/blk_non_zero/TEST5.ps1:39: on their outpute
> src/test/blk_non_zero/TEST6.ps1:39: on their outpute
> src/test/blk_non_zero/TEST7.ps1:39: on their outpute
> src/test/blk_non_zero/TEST8.ps1:39: on their outpute
> src/test/blk_non_zero/TEST9.ps1:39: on their outpute</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[src/common/out.c, line 243 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuMgyCHh-4VjEkwalw:-KI2vIp3XMkutWyZ7GrG:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/common/out.c#L243)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> Sure, I will try that thanks!
</details>
Done.

---

*[src/test/RUNTESTS, line 107 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv8nUEkl5zgABaVrmo:-KI2vL_FdeOBn-HOZKFw:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/RUNTESTS#L107)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> OK, didn't know how to do that :)
</details>
Done.

---

*[src/windows/include/platform.h, line 71 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuXD3fW07fEExdzUAw:-KI2vYRoXX4dvl2z1xG7:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L71)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> it is used somewhere, I will remove and find out then replace (I added this because of failure to compile)
</details>
Done.

---

*[src/windows/include/platform.h, line 79 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuXUn8yOOHS3FzW45D:-KI2vyG0PUCbW4T9h-g9:-572010044) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L79)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> Again, I added it because of a failure to compile so something got missed somewhere.  I'll take it out and see who complains
</details>
unittest.h has a bunch of these in it, for example:
int ut_dirfd(const char *file, int line, const char *func, DIR *dirp);

---

*[src/windows/include/platform.h, line 81 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuplINV4slhbE_QNwM:-KI2wUfYxPJ5E-eJ5xiI:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L81)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> will take a look
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[src/test/blk_nblock/blk_nblock.c, line 65 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHux6C-jMc4eHq5ZKeZ:-KI40hnbb0M9dygRUcBw:1102673173) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_nblock/blk_nblock.c#L65)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> OK, I'll go look and see what this is but if you add some more detail here that's be great.  These new macros were introduced after I started and I still had more work to do so switched to the ones that we were using that don't casus e VS to throw up
</details>
I'm going to punt on this one. I'm not sure how to fix and its not critical in light of the need for this patch to land. Either shoot me some specifics on what you were thinking or knock it out on that end :)

---

*[src/test/unittest/unittest.ps1, line 40 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv5OXRaUr37pPHxvbY:-KI41YkgCWu71YkDnl5p:-423517852) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/unittest/unittest.ps1#L40)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> yeah, my editor was setup for spaces instead of tabs at first so everything I touched got converted. Then a bunch of tests failed so I changed it to use tabs instead of spaces and remove trailing white space but I need to touch any files that are affected.  What is our tab count?
</details>
this may not be 100% across all windows files added since the start, we'll have to fix them as we find them I think

---

*[src/windows/include/platform.h, line 78 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHueUBzF_iquaI-ymAb:-KI421SXL8MX8rmkuTDL:1676949375) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L78)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> which one?  If I added it, it was required to compile (unless copy n paste error or something).  If not then it was already there
</details>
they seem needed to me, we can remove in a followon patch

---

*[src/windows/include/platform.h, line 79 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuXUn8yOOHS3FzW45D:-KI427DNAvyv-4TQUa6T:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L79)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> unittest.h has a bunch of these in it, for example:
> int ut_dirfd(const char *file, int line, const char *func, DIR *dirp);
</details>
Done.

---

*[src/windows/include/platform.h, line 80 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuUC9Ae0WdodgvGufD:-KI428jfsWp4zNXqiDFO:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L80)):*
<details><summary><i>Previously, krzycz (Krzysztof Czurylo) wrote…</i></summary>
> See comment in out.c
</details>
Done.

---

*[src/windows/include/platform.h, line 81 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHuplINV4slhbE_QNwM:-KI42AmWfrjJzWkTAzhF:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L81)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> Done.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[src/test/blk_nblock/blk_nblock.c, line 65 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHux6C-jMc4eHq5ZKeZ:-KI42Fde1p1oVQrUkwzz:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/blk_nblock/blk_nblock.c#L65)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> I'm going to punt on this one. I'm not sure how to fix and its not critical in light of the need for this patch to land. Either shoot me some specifics on what you were thinking or knock it out on that end :)
</details>
Done.

---

*[src/windows/include/platform.h, line 78 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHueUBzF_iquaI-ymAb:-KI42Kdzkj-Dvxzv5OTG:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/windows/include/platform.h#L78)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> they seem needed to me, we can remove in a followon patch
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: 57 of 101 files reviewed at latest revision, 4 unresolved discussions.

---

*[src/test/unittest/unittest.ps1, line 40 \[r1\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KHv5OXRaUr37pPHxvbY:-KI85SPYBnJKezywLnOo:-499146009) ([raw file](https://github.com/pmem/nvml/blob/d807b3a3ace335b13cdf2539a539150d0b00cc54/src/test/unittest/unittest.ps1#L40)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> this may not be 100% across all windows files added since the start, we'll have to fix them as we find them I think
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 57 of 101 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/common/util.c, line 554 \[r5\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KI8Z_A_y_pyRkclK-ko:-KI8Z_A_y_pyRkclK-kp:-1456329033) ([raw file](https://github.com/pmem/nvml/blob/8e334368138fced5daf26b64ad8bb997bf409604/src/common/util.c#L554)):*
> ```C
> 		return -1;
> 	}
> #endif
> ```

Please make a variable for mode and initialize it based on platform.

---

*[src/libpmemobj/memblock.c, line 87 \[r5\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KI8_SY6LvG6wuL-qMun:-KI8_SY7jLYlXct1P-zs:579091249) ([raw file](https://github.com/pmem/nvml/blob/8e334368138fced5daf26b64ad8bb997bf409604/src/libpmemobj/memblock.c#L87)):*
> ```C
> 			FATAL("possible zone chunks metadata corruption");
> 	}
> 	return retValue;
> ```

This change suggests FATAL "unreachable" annotation does not work for your compiler. Please fix it if possible and don't change this code.

---

*[src/test/testconfig.ps1.example, line 14 \[r5\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KI8_jpDhA1wYU08oiCc:-KI8_jpDhA1wYU08oiCd:1433195674) ([raw file](https://github.com/pmem/nvml/blob/8e334368138fced5daf26b64ad8bb997bf409604/src/test/testconfig.ps1.example#L14)):*
> ```
> # elsewhere, set the path here.
> #
> sv -name LOCAL_FS_DIR '.'
> ```

LOCAL_FS_DIR does not exist anymore. Please don't use it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 57 of 101 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/libpmemobj/memblock.c, line 87 \[r5\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KI8_SY6LvG6wuL-qMun:-KI8bQ9B-Ldse10t52Kf:-2000955550) ([raw file](https://github.com/pmem/nvml/blob/8e334368138fced5daf26b64ad8bb997bf409604/src/libpmemobj/memblock.c#L87)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> This change suggests FATAL "unreachable" annotation does not work for your compiler. Please fix it if possible and don't change this code.
</details>
s/unreachable/noreturn/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: 57 of 101 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/libpmemobj/memblock.c, line 87 \[r5\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KI8_SY6LvG6wuL-qMun:-KI8c0XanTIufRouKGS0:-2145234251) ([raw file](https://github.com/pmem/nvml/blob/8e334368138fced5daf26b64ad8bb997bf409604/src/libpmemobj/memblock.c#L87)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> s/unreachable/noreturn/
</details>
No, the problem is returning from within a case statement, no all control paths possible. What is your issue with the change?

---

*[src/test/testconfig.ps1.example, line 14 \[r5\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KI8_jpDhA1wYU08oiCc:-KI8cCLXw4U7s74zJavq:1315489394) ([raw file](https://github.com/pmem/nvml/blob/8e334368138fced5daf26b64ad8bb997bf409604/src/test/testconfig.ps1.example#L14)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> LOCAL_FS_DIR does not exist anymore. Please don't use it. 
</details>
OK, will fix this in an upcoming patch. Windows stuff is not production yet, we are trying to get UT landed so a host of other developers can begin to contribute

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

peluse commented:
Review status: 57 of 101 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/common/util.c, line 554 \[r5\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KI8Z_A_y_pyRkclK-ko:-KI8ebZmd82AJRP9KY5_:-1874643704) ([raw file](https://github.com/pmem/nvml/blob/8e334368138fced5daf26b64ad8bb997bf409604/src/common/util.c#L554)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary>
> Please make a variable for mode and initialize it based on platform.
</details>
Please be more specific, not sure what you mean.  Also, this would be addressed in a follow-on patch.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

marcinslusarz commented:
Review status: 57 of 101 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/common/util.c, line 554 \[r5\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KI8Z_A_y_pyRkclK-ko:-KI8f3RbNtM3A1XOiMb9:302466638) ([raw file](https://github.com/pmem/nvml/blob/8e334368138fced5daf26b64ad8bb997bf409604/src/common/util.c#L554)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> Please be more specific, not sure what you mean.  Also, this would be addressed in a follow-on patch.
</details>
int flags;
#ifdef _WIN32
  flags = S_IWRITE | S_IREAD;
#else
  flags = 0;
#endif

if ((fd = open(...)))
....

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->

krzycz commented:
Merged by hand.
Remaining issues/comments will be addressed in a separate patch.

marcinslusarz commented:
*[src/libpmemobj/memblock.c, line 87 \[r5\]](https://reviewable.io:443/reviews/pmem/nvml/819#-KI8_SY6LvG6wuL-qMun:-KI8flRYw_NrrtdMxg8e:-412462506) ([raw file](https://github.com/pmem/nvml/blob/8e334368138fced5daf26b64ad8bb997bf409604/src/libpmemobj/memblock.c#L87)):*
<details><summary><i>Previously, peluse (Paul Luse) wrote…</i></summary>
> No, the problem is returning from within a case statement, no all control paths possible. What is your issue with the change?
</details>
This change is not needed. The problem is with the definition of FATAL macro on Windows.
Without attribute((noreturn)) on out_fatal this would also not compile on Linux.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pmem/nvml/819)*
<!-- Sent from Reviewable.io -->
